### DISCUSSION TITLE
Cannot set remote user in v2.4.5+

### DISCUSSION THREAD

**Discussion by jlbooker** [OpenIDC/mod_auth_openidc#544] [Q&A] [answered]
Hi there,

I'm currently using mod_auth_openidc v2.4.4. I was looking to upgrade to v2.4.6 because we're running into the semaphore cleanup on graceful restarts bug that's fixed in the latest release.

However, upon upgrading to v2.4.6, I now get the following error when trying to login against MS Azure:
`OpenID Connect Provider error: Remote user could not be set: contact the website administrator`

The Apache configs didn't change at all during the upgrade, I just replaced the mod_auth_opendic.so module file. I downgraded to v2.4.4 and the remote user error went away.

I then tried upgrading to v2.4.5 and received the same error regarding setting the remote user, so it seems to be a change between 2.4.4 and 2.4.5.

In the Apache logs, I see:
```
[auth_openidc:error] [pid 11879] [client x] oidc_proto_resolve_userinfo: mandatory claim ("sub") was not returned from userinfo endpoint (https://openid.net/specs/openid-connect-core-1_0.html#UserInfoResponse), referer: https://fedauth.mycompany.com/
[auth_openidc:error] [pid 11879] [client x] oidc_retrieve_claims_from_userinfo_endpoint: resolving user info claims with the existing/provided access token failed, nothing will be stored in the session, referer: https://fedauth.mycompany.com/
[auth_openidc:warn] [pid 11879] [client x] oidc_get_remote_user: JSON object did not contain a "onPremisesSamAccountName" string, referer: https://fedauth.mycompany.com/
[auth_openidc:error] [pid 11879] [client x] oidc_set_request_user: OIDCRemoteUserClaimis set to "onPremisesSamAccountName", but could not set the remote user based on the requested claim "onPremisesSamAccountName" and the available claims for the user, referer: https://fedauth.mycompany.com/
[auth_openidc:error] [pid 11879] [client x] oidc_handle_authorization_response: remote user could not be set, referer: https://fedauth.mycompany.com/
[auth_openidc:error] [pid 11879] [client x]  oidc_restore_proto_state: no "mod_auth_openidc_state_9tTSRXjUM8sB6DGCNUUGZrYEfyc" state cookie found, referer: https://mycompany.com/secure/redirect_uri?code=[long token here]
[auth_openidc:error] [pid 11879] [client x] oidc_unsolicited_proto_state: could not parse JWT from state: invalid unsolicited response: [src/jose.c:809: oidc_jwt_parse]: cjose_jws_import failed: invalid argument [file: jws.c, function: cjose_jws_import, line: 787], referer: https://mycompany.com/secure/redirect_uri?code=[long token here]
```

Relevant Apache config is:
```
OIDCProviderJwksUri https://login.microsoftonline.com/<app id here>/discovery/v2.0/keys
OIDCProviderUserInfoEndpoint "https://graph.microsoft.com/v1.0/me?$select=mailNickname,onPremisesSamAccountName"
OIDCRemoteUserClaim onPremisesSamAccountName
```

**Comment by zandbelt** [comment]
there's a security improvement in 2.4.5 indeed, affecting the validation of the results returned from the userinfo endpoint, see: https://github.com/zmartzone/mod_auth_openidc/commit/4d943280daa5f0866b50c6dfcbe295e56608bf10 ; however, for Azure AD I don't think you should be setting individual URLs for the endpoints as you did but rather follow  https://github.com/zmartzone/mod_auth_openidc/wiki/Azure-Active-Directory-Authentication and set `OIDCProviderMetadataURL https://sts.windows.net/[TENANT_ID]/.well-known/openid-configuration`

**Reply by jlbooker** [reply]
We do set a `OIDCProviderMetadataURL` too, but I believe we also needed to set the `OIDCProviderUserInfoEndpoint` because the `mailNickname` field isn't included in the default metadata.

**Reply by jlbooker** [reply]
Testing without the `UserInfoEndpoint` on v2.4.4 cleans up some of the errors, but still shows:
```
oidc_get_remote_user: JSON object did not contain a "mailNickname" string
oidc_set_request_user: OIDCRemoteUserClaimis set to "mailNickname", but could not set the remote user based on the requested claim "mailNickname" and the available claims for the user
```

**Reply by jlbooker** [reply]
I've also tried `onPremisesSamAccountName` for the `OIDCRemoteUserClaim`, with the same result.

**Comment by zandbelt** [comment, marked as answer]
you cannot use `mailNickname` as `OIDCRemoteUserClaim` but you'll need to pick a claim that is actually returned from the user info endpoint; calling another endpoint that returns no "sub" claim is just not part of OpenID Connect, not supported by the module and insecure because of known attacks

**Reply by jlbooker** [reply]
Hmm, it appears I'm stuck on v2.4.4 then. According to the Azure docs, the [UserInfo response cannot be customized](https://docs.microsoft.com/en-us/azure/active-directory/develop/userinfo#userinfo-response). It only returns bare-bones information -- no username, upn, sAMAccountName, etc. It seems others have had [the same question](https://social.msdn.microsoft.com/Forums/en-US/eb448c25-4630-4e2e-9d0b-7db8b7ab2e72/azure-ad-openid-connect-v2-userinfo-not-returning-username-etc?forum=WindowsAzureAD).

It seems like Azure expects apps to use their graph endpoints (like https://graph.microsoft.com/v1.0/me ) for additional information, but as you pointed out, those endpoints are not compliant with the oidc spec.

I hate to ask.. But is there any way to add an option to mod_auth_openidc to allow the previous (less secure, non-spec) functionality to continue working? I don't see how anyone can use this with Azure, with the UserInfo endpoint being so sparse. I'm now doing lots of mitigation work to keep Apache from running out of semaphores and crashing our sites, but I can't upgrade to the patched version of this module. It's a bad state to be stuck in.

**Reply by zandbelt** [reply]
let me think this over a bit: it's awkward indeed to add a non-compliant "feature" that is a known vulnerability

**Reply by zandbelt** [reply]
I created a - rather undocumented - workaround to maintain backwards compatibility, see 5e21d955a816f513a99378429b65f5a866fdf381; you should be able to use something like `SetEnvIfExpr true OIDC_NO_USERINFO_SUB =true` in the Apache config to trigger it

**Reply by jlbooker** [reply]
Thanks so much @zandbelt ! Do you see the next release going out sometime soon, or should I set about building from source to grab this update?

**Reply by zandbelt** [reply]
My guess is that a new release is more than just a few weeks out.

**Reply by jlbooker** [reply]
I've built and installed `mod_auth_openidc-2.4.7-dev.so`. The semaphore issue is confirmed fixed (yay!), so thanks again. 

I added the `SetEnvIfExpr ...` line to the Apache config. However, I'm now seeing the following errors upon logging in:

```
oidc_util_json_string_print: oidc_util_check_json_error: response contained an "error" entry with value: ""invalid_grant""
oidc_util_json_string_print: oidc_util_check_json_error: response contained an "error_description" entry with value: ""AADSTS54005: OAuth2 Authorization code was already redeemed, please retry with a new valid code or use an existing refresh token.
oidc_proto_resolve_code_and_validate_response: failed to resolve the code
```
The browser lands on the redirect_uri with a message that says:
```
OpenID Connect Provider error: Error in handling response type.
```

The openidc responses from Azure haven't changed, and the Apache configs only have the one new line.

**Reply by zandbelt** [reply]
I see no link to that latest change. I'm pretty sure that just unsetting/setting `SetEnvIfExpr` would yield the same results, please confirm; if it doesn't, I'm interested in the server debug logs; also, make sure to run this in a fresh browser or incognito tab

**Reply by jlbooker** [reply]
Removing the `SetEnvIfExpr` line took me back to the previous error:

Browser says: `OpenID Connect Provider error: Remote user could not be set: contact the website administrator`

Apache error log has:
```
oidc_proto_resolve_userinfo: mandatory claim ("sub") was not returned from userinfo endpoint (https://openid.net/specs/openid-connect-core-1_0.html#UserInfoResponse)
oidc_retrieve_claims_from_userinfo_endpoint: resolving user info claims with the existing/provided access token failed, nothing will be stored in the session
oidc_get_remote_user: JSON object did not contain a "mailNickname" string
oidc_handle_authorization_response: remote user could not be set
```

And yes, I'm testing this in a fresh incognito tab every time.

**Reply by zandbelt** [reply]
I need to see the server debug logs for the case where `SetEnvIfExpr` is set, more than just a few lines, before and after

**Reply by jlbooker** [reply]
Have a look at: [gist removed]

**Reply by zandbelt** [reply]
ok, thanks; the 2 latest commits 4b2044e758f7ad79af88b23bb89b6273f06ef633 should fix this

**Reply by jlbooker** [reply]
Thanks again! This latest build appears to be working well so far!
